widget/list.go

```diff
@@ -0,0 +1,413 @@
+	visibleItemCount int
+	firstItemIndex   int
+	lastItemIndex    int
+}
+
+func (l *listRenderer) BackgroundColor() color.Color {
+	if l.list.background == nil {
+		return theme.BackgroundColor()
+	}
+
+	return l.list.background
+}
+
+func (l *listRenderer) Layout(size fyne.Size) {
+	if l.list.Length() == 0 {
+		return
+	}
+	if l.itemCache == nil {
+		l.itemCache = &listCache{}
+	}
+
+	offsetChange := l.list.previousOffsetY - l.list.offsetY
+
+	if offsetChange < 0 {
+		// Scrolling Down
+		itemChange := 0
+		layoutEndY := l.children[len(l.children)-1].Position().Y + l.list.itemMin.Height
+		scrollerEndY := l.scroller.Offset.Y + l.scroller.Size().Height
+		if layoutEndY < scrollerEndY {
+			itemChange = int(math.Ceil(float64(scrollerEndY-layoutEndY) / float64(l.list.itemMin.Height)))
+		} else if offsetChange*-1 < l.list.itemMin.Height {
+			return
+		} else {
+			itemChange = int(math.Floor(float64(offsetChange*-1) / float64(l.list.itemMin.Height)))
+		}
+		l.list.previousOffsetY = l.list.offsetY
+		for i := 0; i < itemChange && l.lastItemIndex != l.list.Length()-1; i++ {
+			l.itemCache.Release(l.children[0])
+			l.children = l.children[1:]
+			l.firstItemIndex++
+			l.lastItemIndex++
+			l.appendItem(l.lastItemIndex)
+		}
+	} else if offsetChange > 0 {
+		// Scrolling Up
+		itemChange := 0
+		layoutStartY := l.children[0].Position().Y
+		if layoutStartY > l.scroller.Offset.Y {
+			itemChange = int(math.Ceil(float64(layoutStartY-l.scroller.Offset.Y) / float64(l.list.itemMin.Height)))
+		} else if offsetChange < l.list.itemMin.Height {
+			return
+		} else {
+			itemChange = int(math.Ceil(float64(offsetChange) / float64(l.list.itemMin.Height)))
+		}
+		l.list.previousOffsetY = l.list.offsetY
+		for i := 0; i < itemChange && l.firstItemIndex != 0; i++ {
+			l.children[len(l.children)-1].Hide()
+			l.itemCache.Release(l.children[len(l.children)-1])
+			l.children = l.children[:len(l.children)-1]
+			l.firstItemIndex--
+			l.lastItemIndex--
+			l.prependItem(l.firstItemIndex)
+		}
+	} else {
+		// Relayout What Is Visible - no scroll change - initial layout or possibly from a resize
+		l.scroller.Resize(size)
+		l.visibleItemCount = int(math.Ceil(float64(l.scroller.size.Height) / float64(l.list.itemMin.Height)))
+		if len(l.children) >= l.visibleItemCount || len(l.children) >= l.list.Length() {
+			return
+		}
+		for i := len(l.children) + l.firstItemIndex; len(l.children) <= l.visibleItemCount && i < l.list.Length(); i++ {
+			l.appendItem(i)
+		}
+		l.layout.Objects = l.children
+		l.layout.Layout.Layout(l.layout.Objects, l.list.itemMin)
+		l.lastItemIndex = l.firstItemIndex + len(l.children) - 1
+
+		cachePoolCount := 5
+		if l.list.Length()-len(l.children) < 5 {
+			cachePoolCount = l.list.Length() - len(l.children)
+		}
+		for i := l.itemCache.Count(); i < cachePoolCount; i++ {
+			// Make sure to keep extra items in the cache
+			item := newListItem(l.list.CreateItem(), nil)
+			l.itemCache.Release(item)
+		}
```

Comment:
How much change in performance do you see without these lines?

Reply:
I would say performance is definitely noticeable visually without these lines.